raiden/utils/migrations/v18_to_v19.py

```diff
@@ -0,0 +1,167 @@
+import json
+
+from web3 import Web3
+
+from raiden.exceptions import InvalidDBData
+from raiden.storage.sqlite import EventRecord, SQLiteStorage, StateChangeRecord
+from raiden.utils.typing import Any, BlockNumber, Dict
+
+SOURCE_VERSION = 18
+TARGET_VERSION = 19
+
+
+class BlockHashCache(object):
+    """A small cache for blocknumber to blockhashes to optimize this migration a bit
+
+    This cache lives only during the v18->v19 migration where numerous RPC calls are
+    expected to be made, many of them probably querying the same block_number -> block_hash.
+    To save RPC calls and make it a bit faster we keep this short-lived cache for the duration
+    of the migration
+    """
+    def __init__(self, web3: Web3):
+        self.web3 = web3
+        self.mapping = {}
+
+    def get(self, block_number: BlockNumber) -> str:
+        """Given a block number returns the hex representation of the blockhash"""
+        if block_number in self.mapping:
+            return self.mapping[block_number]
+
+        block_hash = self.web3.eth.getBlock(block_number)['hash']
+        block_hash = block_hash.hex()
+        self.mapping[block_number] = block_hash
+        return block_hash
+
+
+def _add_blockhash_to_state_changes(storage: SQLiteStorage, cache: BlockHashCache) -> None:
+    """Adds blockhash to ContractReceiveXXX and ActionInitChain state changes"""
+    state_changes = storage.batch_query_state_changes(batch_size=500)
```

Comment:
If I understand correctly, you didn't really manage to avoid having whole table in-memory, you only get it from db in chuncks but still everything is in-memory at same time.

Reply:
That is correct. It's not finished yet.

Reply:
label got me, sorry for rushing it. I just love db optimizations and got too excited :)

Reply:
No worries my friend. It's good to have more eyes on the code.